BossMod/Modules/RealmReborn/Dungeon/D18TheKeeperoftheLake/D183Midgardsormr.cs

```diff
@@ -0,0 +1,125 @@
+    MirageAdmonishment = 29290, // MirageDragonB/MirageDragonA->self, 2.0s cast, range 40 width 12 rect // done
+    PhantomAdmonishment = 29280, // Boss->self, 7.0s cast, range 40 width 12 rect // done
+    PhantomInnerTurmoil = 29278, // Boss->self, 7.0s cast, range 22 circle // done
+    PhantomKin = 29277, // Boss->self, 4.0s cast, single-target // summons the clones on the side for phantom attacks
+    PhantomOuterTurmoil = 29279, // Boss->self, 7.0s cast, range 39 ?-degree cone
+    UnknownAbility = 30226, // 392D/392E->self, no cast, single-target
+}
+
+class Adonishment(BossModule module) : Components.SelfTargetedAOEs(module, ActionID.MakeSpell(AID.Admonishment), new AOEShapeRect(40, 6));
+class MidgarAdds(BossModule module) : Components.AddsMulti(module, [(uint)OID.MirageDragonC, (uint)OID.MirageDragonD]);
+class Animadversion(BossModule module) : Components.RaidwideCast(module, ActionID.MakeSpell(AID.Animadversion), "Raidwide");
+class Condescension(BossModule module) : Components.SingleTargetCast(module, ActionID.MakeSpell(AID.Condescension)); // tank buster
+class Disgust(BossModule module) : Components.RaidwideCast(module, ActionID.MakeSpell(AID.Disgust), "Raidwide");
+class InnerTurmoil(BossModule module) : Components.SelfTargetedAOEs(module, ActionID.MakeSpell(AID.InnerTurmoil), new AOEShapeCircle(22));
+class MirageAdonishment(BossModule module) : Components.SelfTargetedAOEs(module, ActionID.MakeSpell(AID.MirageAdmonishment), new AOEShapeRect(40, 6));
+class PhandomAdomishment(BossModule module) : Components.SelfTargetedAOEs(module, ActionID.MakeSpell(AID.PhantomAdmonishment), new AOEShapeRect(40, 6));
+class PhantomInner(BossModule module) : Components.SelfTargetedAOEs(module, ActionID.MakeSpell(AID.PhantomInnerTurmoil), new AOEShapeCircle(22));
+class PhantomOuter(BossModule module) : Components.SelfTargetedAOEs(module, ActionID.MakeSpell(AID.PhantomOuterTurmoil), new AOEShapeDonutSector(22, 40, 90.Degrees()));
+
+class Anitpathy(BossModule module) : Components.ConcentricAOEs(module, [new AOEShapeCircle(6), new AOEShapeDonut(6, 12), new AOEShapeDonut(12, 20)])
+{
+    public override void OnCastStarted(Actor caster, ActorCastInfo spell)
+    {
+        if ((AID)spell.Action.ID == AID.AntipathyCast1)
+            AddSequence(Module.Center, Module.CastFinishAt(spell));
+    }
+
+    public override void OnEventCast(Actor caster, ActorCastEvent spell)
+    {
+        var order = (AID)spell.Action.ID switch
+        {
+            AID.AntipathyCast1 => 0,
+            AID.AntipathyCast2 => 1,
+            AID.AntipathyCast3 => 2,
+            _ => -1
+        };
+        if (!AdvanceSequence(order, caster.Position, WorldState.FutureTime(2f)))
+            ReportError($"unexpected order {order}");
+    }
+}
+
+class AhkMornInitial(BossModule module) : Components.StackWithIcon(module, (uint)IconID.AhkMornStack, ActionID.MakeSpell(AID.AkhMornCast), 6, 0, 2, 4);
```

Comment:
i think this whole component can be removed, right? since the AhkMornFollow component also watches for the icon addition. right now we get two "Stack!" hints in the vbm window